core/state_processor.go

```diff
@@ -78,12 +90,18 @@ func (p *StateProcessor) Process(block *types.Block, statedb *state.StateDB, cfg
 	)
 	// Iterate over and process the individual transactions
 	for i, tx := range block.Transactions() {
-		msg, err := TransactionToMessage(tx, signer, header.BaseFee)
+		var msg *Message
+		msg, err = TransactionToMessage(tx, signer, header.BaseFee)
 		if err != nil {
 			return nil, nil, 0, fmt.Errorf("could not apply tx %d [%v]: %w", i, tx.Hash().Hex(), err)
 		}
 		statedb.SetTxContext(tx.Hash(), i)
-		receipt, err := applyTransaction(msg, p.config, gp, statedb, blockNumber, blockHash, tx, usedGas, vmenv)
+
+		var receipt *types.Receipt
+		receipt, err = applyTransaction(msg, p.config, gp, statedb, blockNumber, blockHash, tx, usedGas, vmenv)
+		if vmenv.Config.Tracer != nil {
+			vmenv.Config.Tracer.CaptureTxEnd(receipt)
+		}
```

Comment:
This `CaptureTxEnd` is in excess, it's already done in a defer within `applyTransaction`:

https://github.com/ethereum/go-ethereum/blob/3cabef4c66575f641858ff5ce63120288f1cab44/core/state_processor.go#L127-L129